reproducibility/codalab/reproduce_results.py

```diff
@@ -0,0 +1,436 @@
+                        )
+                        results_dfs = load_results(
+                            f"https://worksheets.codalab.org/rest/bundles/{uuid}/contents/blob",
+                            splits=["val", "test"],
+                            include_in_distribution=True,
+                        )
+                        dataset_results[algorithm].append(results_dfs)
+                else:
+                    # 10 different seeds were used for camelyon17. 3 different seeds for other datasets.
+                    seeds = range(0, 10) if dataset == "camelyon17" else [0, 1, 2]
+                    for seed in seeds:
+                        bundle_name = (
+                            f"{dataset}_{algorithm}_seed{seed}"
+                            if not in_distribution
+                            else f"hp_id_{algorithm}_{dataset}_seed{seed}"
+                        )
+                        uuid = self._get_bundle_uuid(bundle_name, worksheet_uuid)
+                        results_dfs = load_results(
+                            f"https://worksheets.codalab.org/rest/bundles/{uuid}/contents/blob",
+                            splits=["val", "test"],
+                            include_in_distribution=True,
+                        )
+                        dataset_results[algorithm].append(results_dfs)
+            compiled_results = compile_results(
+                dataset, dataset_results, in_distribution
+            )
+
+            subdirectory = "main" if not in_distribution else "id_val"
+            with open(
+                f"reproducibility/codalab/output/{subdirectory}/{dataset}_result.json",
+                "w",
+            ) as f:
+                json.dump(compiled_results, f, indent=4)
+
+    def tune_hyperparameters_id(self):
+        def get_grid(params):
+            grid = []
+            for p in product(*params):
+                grid.append(p)
+            return grid
+
+        worksheet_uuid = self._set_worksheet()
+        datasets_uuids = self._get_datasets_uuids(worksheet_uuid)
+        wilds_src_uuid = self._get_bundle_uuid("wilds", worksheet_uuid)
+
+        for dataset, dataset_uuid in datasets_uuids.items():
+            # The following datasets have in-distribution val sets and has been hyperparameter tuned
+            if dataset not in ID_HYPERPARAMETER_SEARCH_SPACE["datasets"]:
+                continue
+
+            hyperparameters = ID_HYPERPARAMETER_SEARCH_SPACE["datasets"][dataset].keys()
+            dataset_fullname = f"{dataset}_{self._wilds_version}"
+            for hyperparameter_values in get_grid(
+                ID_HYPERPARAMETER_SEARCH_SPACE["datasets"][dataset].values()
+            ):
+                hyperparameter_result = dict()
+                for i, hyperparameter in enumerate(hyperparameters):
+                    hyperparameter_result[hyperparameter] = hyperparameter_values[i]
+                print(hyperparameter_result)
+                self._run(
```

Comment:
Do you think we could factor out all of these run commands, since they're pretty similar?

Reply:
That sounds good to me. I will do this before checking in.